### PR TITLE
Allow "page" route parameter to be used in Mvc controllers

### LINKED ISSUE #6660: .NET Core 2.0 route configuration being ignored
I have a website that was working perfectly fine in .NET core 1.1.

Today -- with the release of .NET Core 2.0 -- I decided to update my application. I changed my Target Framework to .NET Core 2.0, and switched over to using the _Microsoft.AspNetCore.All_ package as outlined here: https://blogs.msdn.microsoft.com/webdev/2017/08/14/announcing-asp-net-core-2-0/

I installed the .NET Core 2.0 Runtime on my server. I restarted my server.

The application builds fine, but when I deploy, it seems my routes -- which worked perfectly in .NET Core 1.1 -- no longer work.

When I try using https://example.com/history (desired URL, which the route SHOULD direct to the "History" action of the "Pages" controller) it complains of a missing View.

If I use https://example.com/pages/history (not desired URL) it loads the home page, not the history page.

Again, this worked fine in .NET Core 1.1

This is my _/Program.cs_:
```
using Microsoft.AspNetCore.Hosting;
using System.IO;

namespace Framework
{
    public class Program
    {
        public static void Main(string[] args)
        {
            var host = new WebHostBuilder()
                .UseKestrel()
                .UseContentRoot(Directory.GetCurrentDirectory())
                .UseIISIntegration()
                .UseStartup<Startup>()
                .Build();

            host.Run();
        }
    }
}
```

This is the portion of my _/Startup.cs_ with the **routes defined at the end**:

```

            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "404",
                    template: "404/{id}",
                    defaults: new { controller = "Pages", action = "Error404"}
                );

                routes.MapRoute(
                    name: "news",
                    template: "news/{page}",
                    defaults: new { controller = "Pages", action = "News", page = 1 }
                );

                routes.MapRoute(
                    name: "history",
                    template: "history/{page=1}/{sortFilter?}/{sortOrder?}",
                    defaults: new { controller = "Pages", action = "History"}
                );

                routes.MapRoute(
                    name: "restoration-resources",
                    template: "restoration-resources/{page=1}/{sortFilter?}/{sortOrder?}",
                    defaults: new { controller = "Pages", action = "RestorationResources"}
                );

                routes.MapRoute(
                    name: "vehicles",
                    template: "vehicles/{vehicle?}/{page=1}/{sortFilter?}/{sortOrder?}",
                    defaults: new { controller = "Pages", action = "Vehicles" }
                );

                routes.MapRoute(
                    name: "gallery",
                    template: "gallery/{gallery?}",
                    defaults: new { controller = "Pages", action = "Gallery" }
                );

                routes.MapRoute(
                    name: "rss",
                    template: "rss/{source?}/{post?}",
                    defaults: new { controller = "Pages", action = "RSS"}
                );

                routes.MapRoute(
                    name: "pages",
                    template: "{action}/{id?}",
                    defaults: new { controller = "Pages", action = "News"}
                );

                routes.MapRoute(
                    name: "default",
                    template: "{controller}/{id?}",
                    defaults: new { controller = "Pages", action = "News"}
                );
            });
        }
    }
}
```

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #6660

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageViewLocationExpander.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ public class PageViewLocationExpander : IViewLocationExpander
     {
         public IEnumerable<string> ExpandViewLocations(ViewLocationExpanderContext context, IEnumerable<string> viewLocations)
         {
-            if (string.IsNullOrEmpty(context.PageName))
+            if (!(context.ActionContext.ActionDescriptor is PageActionDescriptor) || string.IsNullOrEmpty(context.PageName))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageViewLocationExpander.cs]: Consider inverting this if to be more readable

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/BasicTests.cs [hunk 2]
```diff
@@ -376,5 +376,31 @@ public static IEnumerable<object[]> HtmlHelperLinkGenerationData
             // Assert
             Assert.Equal(HttpStatusCode.NotFound, response.StatusCode);
         }
+
+        [Fact]
+        public async Task UsingPageRouteParameterInConventionRouteWorks()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/BasicTests.cs]: conventional

**Comment by rynowak** [comment]
Do these places need to change as well?
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngine.cs#L244
https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngine.cs#L321

**Comment by pranavkm** [comment]
They don't as yet. If we do fix it in the View Engine, would we do the same thing for the inverse case, i.e. not read `controller` or `action` when working with a page?

**Comment by rynowak** [comment]
Maybe....

**Comment by pranavkm** [comment]
The layering makes it tricky to determine if you're working with a `PageActionDescriptor`. I guess we could solve it using a contract `IRequiredRouteValueProvider` but for the 2.0.1 release, the smallest amount of code we could write would be to test if we're not in a `ControllerActionDescriptor`

**Comment by rynowak** [comment]
I would look at `ActionDescriptor.RouteValues` for the presence of `page` vs `controller`.

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngine.cs [hunk 3]
```diff
@@ -240,7 +240,13 @@ private ViewLocationCacheResult LocatePageFromPath(string executingFilePath, str
         {
             var controllerName = GetNormalizedRouteValue(actionContext, ControllerKey);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngine.cs]: leave this alone for 2.0.1?

File: src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngine.cs [hunk 4]
```diff
@@ -240,7 +240,13 @@ private ViewLocationCacheResult LocatePageFromPath(string executingFilePath, str
...
             var areaName = GetNormalizedRouteValue(actionContext, AreaKey);
-            var razorPageName = GetNormalizedRouteValue(actionContext, PageKey);
+            string razorPageName = null;
+            if (actionContext.ActionDescriptor.RouteValues.ContainsKey(PageKey))
+            {
+                // Only calculate the Razor Page name if "page" is registered in RouteValues.
+                razorPageName = GetNormalizedRouteValue(actionContext, PageKey);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngine.cs]: Ideally I'd like to change how `GetNormalizedRouteValue` works to verify it returns a value only if it's present in `RouteValues` but that would be too wide a change for 2.0.1. We could consider it for 2.1.0 though.

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/BasicTests.cs (hunk 2 above)
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/BasicTests.cs]: Will fix it

File: src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngine.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngine.cs]: Sure

**Pull request closed without merging**